vtrace/envitools.py

```diff
@@ -25,36 +36,134 @@ def cmpRegs(emu, trace):
     return True
 
 
-def emuFromTrace(trace):
-    '''
-    Produce an envi emulator for this tracer object.
-    '''
-    arch = trace.getMeta('Architecture')
-    plat = trace.getMeta('Platform')
-    amod = envi.getArchModule(arch)
-    emu = amod.getEmulator()
+skip_mem_opcodes_by_arch = {
+        'i386': (e_i386const.INS_NOP, e_i386const.INS_LEA),
+        'amd64': (e_i386const.INS_NOP, e_i386const.INS_LEA),
+        }
+
+
+class LockstepEmulator:
+    def __init__(self, trace):
+        self.trace = trace
+        self.emu = vutil.emuFromTrace(trace)    # this should set up the emulator
+        self.arch = self.trace.getMeta("Architecture")
+
+    def go(self):
+        count = 0
+        skip_mem_opcodes = skip_mem_opcodes_by_arch.get(self.arch, ())
+
+        self.cmpRegs()
+        while True:
+            print("%4d  Lockstep: 0x%.8x" % (count, self.emu.getProgramCounter()), end='')
+            try:
+                pc = self.trace.getProgramCounter()
+                op = self.emu.parseOpcode(pc)
+                print("  %r  " % op, end='')
+
+                # don't compare memory from these instruction results (error prone)
+                skip_mem = False
+                if op.opcode in skip_mem_opcodes:
+                    skip_mem = True
+
+                # store the current opcode addresses for comparison later (for when the 
+                #  original register gets clobbered
+                if not skip_mem:
+                    self.grabMemAddrs(op)
+
+                for oper in op.opers:
+                    if skip_mem:
+                        continue
+                    print("(%r = %x)" % (oper.repr(op), oper.getOperValue(op, self.emu)), end='')
+
+                self.trace.stepi()
+                self.emu.stepi()
+                count += 1
+
+                if op.mnem == 'rdtsc':
```

Comment:
Hmm. It feels like we should copy the concept of a pre/post hook from our emulator class into here, since rdtsc is specifc to intel

Reply:
agreed.  or at least a architecture-specific handler system.  rdtsc reads from the system clock, which we don't do.  so this triggers a register-resync.